align to jargon version 4.3.7.0 (including updates for spring framework)

jaksat0202 opened the pull request:
Hello,

As promised during the latest UGM, here are small changes to iDrop so that the code compiles against the latest version of jargon (4.3.7.0)

// Jakob Saternus

korydraughn commented:
I tried to compile your PR but ran into the following errors.
```
[INFO] ------------------------------------------------------------------------
[INFO] Reactor Summary for iRODS iDrop 4.3.7.0-RELEASE:
[INFO] 
[INFO] iRODS iDrop ........................................ SUCCESS [  0.149 s]
[INFO] Jargon Conveyor .................................... FAILURE [ 47.776 s]
[INFO] idrop-swing ........................................ SKIPPED
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  48.064 s
[INFO] Finished at: 2025-07-08T16:57:01-04:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project jargon-conveyor: Could not collect dependencies for project org.irods.jargon:jargon-conveyor:jar:4.3.7.0-RELEASE
[ERROR] Failed to read artifact descriptor for org.netbeans.api:org-openide-util:jar:RELEASE801
[ERROR] 	Caused by: The following artifacts could not be resolved: org.netbeans.api:org-openide-util:pom:RELEASE801 (absent): Could not transfer artifact org.netbeans.api:org-openide-util:pom:RELEASE801 from/to netbeans.repository (https://bits.netbeans.org/maven2/): status code: 401, reason phrase: Repository decommissioned. Please refer to https://netbeans.apache.org/about/oracle-transition.html for more information. (401)
[ERROR] 
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <args> -rf :jargon-conveyor
```

Please confirm you're able to build the project from scratch.

jaksat0202 commented:
There was indeed a reference to an old repository (bits.netbeans.org) which is no longer available.
It is fixed in commit 96c38b478a74deb1019989db4b6994cc94be3300.

korydraughn commented:
Your latest commit fixes the build, but the application crashes due to a null pointer exception.
```
[main] INFO  systraygui.iDrop  - building tree to look at staging resource
[main] INFO  systraygui.iDrop  - building tree to look at local file system
[main] ERROR systraygui.IDROPDesktop  - unable to start application due to error
java.lang.NullPointerException: Cannot invoke "java.awt.CheckboxMenuItem.getState()" because "this.pausedItem" is null
	at org.irods.jargon.idrop.desktop.systraygui.iDrop.showIdropGui(iDrop.java:130)
	at org.irods.jargon.idrop.desktop.systraygui.iDrop.signalIdropCoreReadyAndSplashComplete(iDrop.java:221)
	at org.irods.jargon.idrop.desktop.systraygui.IDROPDesktop.doStartupSequence(IDROPDesktop.java:210)
	at org.irods.jargon.idrop.desktop.systraygui.IDROPDesktop.main(IDROPDesktop.java:362)
```

Is the application working for you?

jaksat0202 commented:
Yes, both on Windows and Linux. Did you get this NPE on Mac by any chance?

korydraughn commented:
Hmm. Not sure why I'm getting an error then. I'm running Ubuntu 20.04.

jaksat0202 commented:
Could you compile and run this simple program and paste here the output:

```
public class TestSystemTray {
    public static void main(String[] args) {
        System.out.println("Java version: " + Runtime.version());
        System.out.println("Tray supported: " + java.awt.SystemTray.isSupported());
    }
}
```

korydraughn commented:
Here's the output.
```
Java version: 17.0.15+6-Ubuntu-0ubuntu120.04
Tray supported: false
```

jaksat0202 commented:
Blame on Ubuntu for removing support for System Tray, iDrop will not run without it.

trel commented:
https://stackoverflow.com/questions/72594274/awt-and-ubuntu-20-the-system-tray-is-not-supported-on-the-current-platform

points to here...

https://bugs.openjdk.org/browse/JDK-8289233

korydraughn commented:
> Blame on Ubuntu for removing support for System Tray, iDrop will not run without it.

You said it runs on Linux for you. What distro did you use to confirm that?

jaksat0202 commented:
Linux Mint
https://linuxmint.com/

trel commented on pom.xml:
```diff
@@ -80,6 +83,7 @@
                 <checksumPolicy>warn</checksumPolicy>
             </snapshots>
         </repository>
+        -->
```
should we just ... remove this stanza rather than comment it out?

jaksat0202 replied:
yes.

korydraughn commented on jargon-conveyor/pom.xml:
```diff
@@ -85,10 +89,12 @@
             <groupId>com.mchange</groupId>
             <artifactId>c3p0</artifactId>
         </dependency>
+<!--
         <dependency>
             <groupId>dom4j</groupId>
             <artifactId>dom4j</artifactId>
         </dependency>
+-->
```
Can remove this.

korydraughn replied:
Not a blocker. We can handle this in a separate PR.

Resolving.

korydraughn commented on pom.xml:
```diff
@@ -253,23 +245,28 @@
             <artifactId>slf4j-log4j12</artifactId>
+            <version>1.6.6</version>
         </dependency>
+<!--
         <dependency>
             <groupId>log4j</groupId>
             <artifactId>log4j</artifactId>
         </dependency>
+-->
```
Why is this commented out? Was it not being used?

Jargon was updated to use log4j 2. We should do the same for this project if possible.

jaksat0202 replied:
It was already imported by slf4j-log4j12

A switch from slf4j to log4j 2 should be possible with a few modifications to the dynamic logging configuration done in IdropConfig. I'll have a look at that.

jaksat0202 replied:
Since jargon-pom is in the protected mode. Could you change the scope for log4j-core to be 'compile' there?
It is effectively blocking the switch to log4j 2 in iDrop at the moment.

korydraughn replied:
Changing the scope in jargon-pom shouldn't be necessary? Metalnx uses log4j 2 without needing that.

Try comparing what you have to the pom.xml in this commit: https://github.com/irods-contrib/metalnx-web/commit/a39485a518e4eced8f965311539e515f02c78026

See the following Metalnx commits too. They may be helpful.
- https://github.com/irods-contrib/metalnx-web/issues/374
- https://github.com/irods-contrib/metalnx-web/issues/393

If you're getting a build error, please post it here.

jaksat0202 replied:
This is what I get with log4j-core included in the runtime scope:

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:2.1:compile (default-compile) on project idrop-swing: Compilation failure: Compilation failure:            
[ERROR] \projects\idrop\idrop-swing\src\main\java\org\irods\jargon\idrop\desktop\systraygui\utils\IdropConfig.java:[8,36] error: package org.apache.logging.log4j.core does not exist    
```

I've rewritten the dynamic configuration of the RollingFileAppender that worked for log4j v1 and the code requires 
the class `org.apache.logging.log4j.core.LoggerContext` (among others)  to be visible to the compiler.

But maybe we can skip the dynamic configuration of the RollingFileAppender and have just a static configuration in log4j.xml. It would make the code simpler and work fine with the runtime scope.

?

korydraughn replied:
Simplifying things doesn't sound bad.

Given your PR is focused on updating Jargon to a later version, let's handle updating log4j in a separate PR. That way, your PR lands sooner than later.

jaksat0202 replied:
I've already made a switch to Log4j v2 in the latest commit 6b12cbe58a25a7031455fd3c2954191185102adf
I've removed the dynamic configuration of the rolling file appender and replaced it with a static configuration because of the above mentioned problems with accessing the LoggerContext class.

korydraughn replied:
Good good.

jaksat0202 commented:
It seems like the best solution for the system tray problem is to use [SystemTray](https://github.com/dorkbox/SystemTray)
However, it is a separate issue which this PR was not meant to solve.

korydraughn commented:
We've confirmed the PR builds and runs on Windows.

Please address any remaining review comments. We'll talk about next steps once that's done.

korydraughn commented:
I have a Windows VM on the way. Once I have that configured, I'll confirm the changes in this PR are working.

korydraughn commented on idrop-swing/src/main/resources/idrop.properties:
```diff
@@ -1,4 +1,4 @@
-idrop.version=2.0.1-RC1
+idrop.version=4.3.7.0-SNAPSHOT
```
What's the motivation behind this being 4.3.7.0-SNAPSHOT instead of something like 3.0.0?

Is bumping to 4.3.7.0-SNAPSHOT a requirement?

korydraughn replied:
Hmm, the 2.0.x version didn't align with the README... which makes your change feel more correct.

korydraughn replied:
The JAR files contain 4.3.7.0-SNAPSHOT in the name and idrop doesn't provide an "About" button providing additional info about the application so I think this version number change is okay.

korydraughn reviewed (commented):
Code changes look good overall.

Will test this today.

korydraughn commented:
I've confirmed the latest version of this PR builds and runs on Windows. This was tested against iRODS 4.3.4 and 5.0.2. Uploads and downloads work.

I think we're ready for squashing.

korydraughn commented:
@jaksat0202 Please squash the commits to one commit and update the commit message to use the following structure.
```
[#142] <brief_description> ...
```

korydraughn commented:
You can include additional information about log4j2 and other changes within the body of the commit message.

jaksat0202 committed:
#142 updated iDrop to use jargon version 4.3.7.0

- moved logger declarations to the top of 5 classes to satisfy Codacy
- switch to log4j v2 with static configuration of rolling file appender
- upgrade of Derby to version 10.16.1
- removed the decommissioned repo at bits.netbeans.org
- align to jargon version 4.3.7.0 (including updates for spring framework)

jaksat0202 commented:
Thanks for the code review.
I've squashed the latest commits into one.

korydraughn commented:
Thanks again for the contribution.

Merging.
